amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpConnectorsSpec.scala

```diff
@@ -284,20 +286,22 @@ class AmqpConnectorsSpec extends AmqpSpec {
       publisher.expectRequest()
       publisher.sendNext(ByteString("five"))
 
+      publisher.sendComplete()
+
       // this should lead to all five being fetched into the buffer
       // but we just consume one before we cancel
       subscriber.request(1)
       subscriber.expectNext().bytes.utf8String shouldEqual "one"
-
       subscriber.cancel()
-      publisher.sendComplete()
 
       val subscriber2 = TestSubscriber.probe[IncomingMessage]()
       amqpSource.addAttributes(Attributes.inputBuffer(1, 1)).runWith(Sink.fromSubscriber(subscriber2))
 
       subscriber2.ensureSubscription()
-      subscriber2.request(4)
-      subscriber2.expectNext().bytes.utf8String shouldEqual "two"
+      subscriber2.request(3)
+      // MapAsync execute the provided function as it pull the elements regardless of whether they are passed
+      // downstream or not. Thus the second element is acked even though it didn't get pushed down by `subscriber`.
```

Comment:
This was true (and racy) before as well (although maybe it did not manifest itself in this test), an element could be fetched from the MQ and ack:ed but have downstream cancel/fail before it was emitted.

Reply:
I commented that line to fix the test in akka 2.4 but broke on akka 2.5.

I think that the issue is that mapAsync use to run the function as soon as it pulls it regardless of whether it was emitted or not. So that test could never work with mapAsync.

However, it seems to work as expected in 2.5. The only change I can see in the history is this: https://github.com/akka/akka/commit/a40826e1cbbcdc6a222f7a468efa3c00305db1fe. Any idea??

Reply:
To be honest, we just renamed this source as `atMostOnceSource` which means precisely this. IF things go south in unexpected ways, you might miss the element.

If you want to be 100% sure that you don't miss any element you should use `committableSource` which could also be `atLeastOnceSource`

Reply:
I take back what I just said 2 comments above. 2.5 Didn't fix mapAsync 🙂 

I guess the test it's just 100% indeterministic since whether the element is acked or not depends completely o how fast the probe close the connection... Still open to suggestions :)